src/Avalonia.Base/Animation/Animation.cs

```diff
@@ -21,155 +21,114 @@ public class Animation : AvaloniaObject, IAnimation
         /// <summary>
         /// Defines the <see cref="Duration"/> property.
         /// </summary>
-        public static readonly DirectProperty<Animation, TimeSpan> DurationProperty =
-            AvaloniaProperty.RegisterDirect<Animation, TimeSpan>(
-                nameof(Duration),
-                o => o._duration,
-                (o, v) => o._duration = v);
+        public static readonly StyledProperty<TimeSpan> DurationProperty = AvaloniaProperty.Register<Animation, TimeSpan>(nameof(Duration));
```

Comment:
Can animation properties be even styled? Not to mention, animated.

Reply:
Maybe not, but the goal here is for every `DirectProperty` to end up read-only.

Reply:
The starting point for these changes was the observation that with the new `ValueStore` in place, a local value goes through pretty much the exact same code path for both styled and direct properties.

I'm not keen on making exceptions to these refactors because I hope at the end of this process that we can mark the `setter` argument of `AvaloniaProperty.RegisterDirect` as obsolete, so that it doesn't start sneaking back in over time.

Reply:
> The starting point for these changes was the observation that with the new ValueStore in place, a local value goes through pretty much the exact same code path for both styled and direct properties.

That's really true: in the case of reading, direct properties are vastly more performant as they're a simple CLR property read inlined to a field read in most cases. I modified some of our benchmarks to get some numbers:

|                             Method |        Mean |     Error |    StdDev | Ratio | RatioSD | Allocated | Alloc Ratio |
|----------------------------------- |------------:|----------:|----------:|------:|--------:|----------:|------------:|
|               GetClrPropertyValues |    375.1 ns |   2.70 ns |   2.52 ns |  1.00 |    0.00 |         - |          NA |
|                   Get_Local_Values | 14,330.4 ns | 168.66 ns | 157.77 ns | 38.21 |    0.49 |         - |          NA |

|          Method |     Mean |     Error |    StdDev | Ratio | RatioSD | Allocated | Alloc Ratio |
|---------------- |---------:|----------:|----------:|------:|--------:|----------:|------------:|
| SetDirectValues | 4.397 us | 0.0274 us | 0.0229 us |  1.00 |    0.00 |         - |          NA |
| SetStyledValues | 6.006 us | 0.0874 us | 0.0730 us |  1.37 |    0.02 |         - |          NA |

You can see by looking at the `RatioSD` column that getting a simple `LocalValue` styled property is nearly 40x slower than reading a CLR property (i.e. a direct property) and setting the same property is 1.37x slower. I'm not sure why the benchmarks don't show it, but setting a styled property also requires allocations whereas direct properties are just stored as fields and so obviously don't.

IMO this is an acceptable hit to ensure that all of our control properties can by styled, but I don't see any reason to take this hit for properties which can't be styled.

Reply:
Huh, I got that completely wrong then. My apologies!

This raises new questions about how we prevent the creation of direct properties that might receive non-local values.

Reply:
My first idea is multiple `RegisterDirect` methods. (Disclaimer: I'm not suggesting changes for this PR, just translating my thoughts into pseduocode in order to more precisely define what I mean.)

* Overload A
  * Does NOT have a `setter` parameter 
  * Is constrained to `TOwner : AvaloniaObject`
* Overload B
  * Has a `setter` parameter
  * Is constrained to `TOwner : UnstyledElement`

Where `UnstyledElement` is an empty class that derives directly from `AvaloniaObject`. Its sole purpose is to be incompatible with `StyledElement`. Types like `Animation` would derive from it.

This restricts writeable direct properties to types which definitely will not receive styles, templates, etc.

Reply:
I'm nervous about properties like Text on TextBlock becoming slower it is not uncommon to have tables of many dynamically changing textblocks, Avalonia performance has made huge strides over the years, It would be a shame to go backwards.

Reply:
Animation does seem like it should remain with direct properties for now. Originally, I was just thinking to update controls. Perhaps this should be done later, if at all, after more discussion. Would especially like to know what advantages in real-world usage there is for these to be styled. I don't recall seeing an issue with these being direct before (unlike control properties) and it seems like a good use case for them.

Concerning TextBlock.Text hopefully that will be its own PR and we can do some benchmarking. Not being able to set Text from styles is very unintuitive. WPF worked fine for this case and it was used.

Reply:
> Where UnstyledElement is an empty class that derives directly from AvaloniaObject. Its sole purpose is to be incompatible with StyledElement. Types like Animation would derive from it.

Not sure we can codify this policy into the API as customers may be using direct properties for their own controls. For example we have one customer which is running on 2005-era embedded hardware and they use direct properties almost everywhere there: after all they fully control the API and so can make the trade-off decision themselves.

Having said that, we can open a discussion about it if you like.

> I'm nervous about properties like Text on TextBlock becoming slower it is not uncommon to have tables of many dynamically changing textblocks, Avalonia performance has made huge strides over the years, It would be a shame to go backwards.

At the worst we can do what WinUI does in lots of places: have a private field to cache the current value, which is updated in `OnPropertyChanged`. That would help with the difference in property read performance anyway.

Reply:
That's your choice ultimately, but the nice thing about analyzers is that they can be freely suppressed or deprioritised by the user. Given the exceptions that can now be throw when setting a `DirectProperty`, the issue is something that this customer definitely wants to be alerted to! Others can disable the message if they please.